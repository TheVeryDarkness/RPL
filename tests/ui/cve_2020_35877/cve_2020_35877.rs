use memmap::{Mmap, Protection};
use std::{
    io::Result,
    marker::PhantomData,
    ops::Index,
    sync::{Arc, Mutex},
};

/// A buffer that is backed by an anonymous memory map.
///
/// **Anonymously** means that the memory is not backed by a file.
pub struct AnonymousBuffer<T>
where
    T: Sized,
{
    data: Arc<Mutex<Mmap>>,
    phantom: PhantomData<T>,
}

impl<T> AnonymousBuffer<T>
where
    T: Sized,
{
    /// Create a new anonymous buffer of the given size.
    pub fn try_new(size: usize) -> Result<Self> {
        let map = Mmap::anonymous(size, Protection::ReadWrite)?;
        Ok(Self {
            data: Arc::new(Mutex::new(map)),
            phantom: PhantomData,
        })
    }
}

impl<T> Index<usize> for AnonymousBuffer<T>
where
    T: Sized,
{
    type Output = T;

    fn index(&self, idx: usize) -> &Self::Output {
        unsafe {
            let mut count = idx;
            let mut p: *const T = self.data.lock().unwrap().ptr() as *const T;
            while count > 0 {
                count -= 1;
                p = p.offset(1);
                //FIXME: ~^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
            }
            &*p
            //~^ERROR: it is unsound to dereference a pointer that is offset using an unchecked integer
        }
    }
}

fn main() {
    let buf = AnonymousBuffer::<u8>::try_new(1024).unwrap();
    println!("{:?}", &buf[1024]);
}
