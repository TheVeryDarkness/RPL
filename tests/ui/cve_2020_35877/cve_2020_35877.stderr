error: it is unsound to dereference a pointer that is offset using an unchecked integer
  --> tests/ui/cve_2020_35877/cve_2020_35877.rs:49:13
   |
LL |     fn index(&self, idx: usize) -> &Self::Output {
   |                     --- offset passed in here
...
LL |             let mut p: *const T = self.data.lock().unwrap().ptr() as *const T;
   |                                   ------------------------------------------- pointer created here
...
LL |             &*p
   |             ^^^ dereferenced here
   |
   = help: check whether it's in bound before dereferencing

error: it is unsound to dereference a pointer that is offset using an unchecked integer
  --> tests/ui/cve_2020_35877/cve_2020_35877.rs:49:13
   |
LL |     fn index(&self, idx: usize) -> &Self::Output {
   |                     --- offset passed in here
...
LL |                 p = p.offset(1);
   |                 --------------- pointer created here
...
LL |             &*p
   |             ^^^ dereferenced here
   |
   = help: check whether it's in bound before dereferencing

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35877/cve_2020_35877.rs:46:23
   |
LL |                 p = p.offset(1);
   |                     - ^^^^^^^^^ offset here
   |                     |
   |                     pointer created here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

error: aborting due to 3 previous errors

